cs/di-extensions.texy

```diff
@@ -5,73 +5,110 @@ DI: Tvorba rozšíření
+
+
+Konfigurace rozšíření
+---------------------
+
+Rozšíření můžeme konfigurovat v sekci, jejíž název je stejný jako ten, pod kterým bylo rozšíření přidáno, tedy *blog*.
 
 /--neon
-blog: # stejné jméno jako má extension
+# stejné jméno jako má extension
+blog:
 	postsPerPage: 10
 	comments: false
 \--
 
-Nastavení si lze přečíst v proměnné `$this->config` v rozšíření.
+K definovaným hodnotám se v rozšíření dostaneme přes proměnnou `$this->config`.
 
 /--php
-class MyBlogExtension extends Nette\DI\CompilerExtension
+class BlogExtension extends Nette\DI\CompilerExtension
 {
 	public function loadConfiguration()
 	{
-		dump($this->config);
-		// [2] [ 'postsPerPage' => 10, 'comments' => false ]
+		$config = $this->config;
+	}
+}
 \--
 
-Protože uznáváme princip [konvence před konfigurací |http://en.wikipedia.org/wiki/Convention_over_configuration], nastavíme si výchozí hodnoty, aby aplikace fungovala i bez toho, že budeme něco nastavovat, a sloučíme je hodnotami z konfiguračního souboru.
+
+Výchozí hodnoty konfigurace
+---------------------------
+
+Pokud to lze, je užitečné podle principu [konvence před konfigurací |http://en.wikipedia.org/wiki/Convention_over_configuration] jednotlivým prvkům konfigurace poskytnout výchozí hodnoty, aby uživatel nemusel konfigurovat úplně vše.
+
+Pro sloučení uživatelské konfigurace a výchozích hodnot můžeme použít metodu `validateConfig()`. Jako parametr jí předáme pole všech povolených klíčů a jejich výchozích hodnot, o které rozšíří `$this->config`. Zároveň pokud se v konfiguraci objeví klíče, které nejsou mezi povolenými, bude vyhozena výjimka. Tak se předejde tomu, že někdo omylem splete název klíče v konfiguračním souboru.
+
+/--neon
+# stejné jméno jako má extension
+blog:
+	postsPerPage: 10
+\--
 
 /--php
-class MyBlogExtension extends Nette\DI\CompilerExtension
+class BlogExtension extends Nette\DI\CompilerExtension
 {
-	public $defaults = [
+
+	private $defaults = [
 		'postsPerPage' => 5,
 		'comments' => true
 	];
 
 	public function loadConfiguration()
 	{
-		$this->config += $this->defaults;
-\--
-
-Kompletní nastavení máme v poli `$this->config`, takže je použijeme pro služby, které vytvoříme. Použijeme k tomu `ContainerBuilder`, který umožňuje to stejné, jako zápis služby v konfiguračním souboru.
-
-/--php
-$builder = $this->getContainerBuilder();
-\--
-
-Vytvoříme si jako příklad službu `articles`, což bude objekt třídy `MyBlog\ArticlesModel`, nad kterým ještě zavoláme metodu `setLogger`.
-
-.[tip]
-Konvence je, prefixovat služby v rozšíření jeho názvem, aby nevznikaly konflikty. Pomůže nám s tím metoda `prefix()`.
-
-/--php
-$builder->addDefinition($this->prefix('articles'))
-	->setFactory('MyBlog\ArticlesModel', ['@connection'])
-	->addSetup('setLogger', ['@logger']);
+		$config = $this->validateConfig($this->defaults);
```

Comment:
Tady nemusí být zřejmé, že to modifikuje i `$this->config`. Co třeba `$config = ` uplně vynechat v následujícím kodu použít `$this->config`?